benchmark/solver/solver.cpp

```diff
@@ -154,61 +219,98 @@ std::shared_ptr<const gko::stop::CriterionFactory> create_criterion(
-                                std::shared_ptr<const gko::Executor>,
-                                std::shared_ptr<const gko::LinOpFactory>)>>
-    solver_factory{{"bicgstab", create_solver<gko::solver::Bicgstab<etype>>},
-                   {"bicg", create_solver<gko::solver::Bicg<etype>>},
-                   {"cg", create_solver<gko::solver::Cg<etype>>},
-                   {"cgs", create_solver<gko::solver::Cgs<etype>>},
-                   {"fcg", create_solver<gko::solver::Fcg<etype>>},
-                   {"idr",
-                    [](std::shared_ptr<const gko::Executor> exec,
-                       std::shared_ptr<const gko::LinOpFactory> precond) {
-                        return gko::solver::Idr<etype>::build()
-                            .with_criteria(create_criterion(exec))
-                            .with_subspace_dim(FLAGS_idr_subspace_dim)
-                            .with_kappa(static_cast<rc_etype>(FLAGS_idr_kappa))
-                            .with_preconditioner(give(precond))
-                            .on(exec);
-                    }},
-                   {"gmres",
-                    [](std::shared_ptr<const gko::Executor> exec,
-                       std::shared_ptr<const gko::LinOpFactory> precond) {
-                        return gko::solver::Gmres<etype>::build()
-                            .with_criteria(create_criterion(exec))
-                            .with_krylov_dim(FLAGS_gmres_restart)
-                            .with_preconditioner(give(precond))
-                            .on(exec);
-                    }},
-                   {"lower_trs",
-                    [](std::shared_ptr<const gko::Executor> exec,
-                       std::shared_ptr<const gko::LinOpFactory>) {
-                        return gko::solver::LowerTrs<etype>::build()
-                            .with_num_rhs(FLAGS_nrhs)
-                            .on(exec);
-                    }},
-                   {"upper_trs",
-                    [](std::shared_ptr<const gko::Executor> exec,
-                       std::shared_ptr<const gko::LinOpFactory>) {
-                        return gko::solver::UpperTrs<etype>::build()
-                            .with_num_rhs(FLAGS_nrhs)
-                            .on(exec);
-                    }},
-                   {"overhead", create_solver<gko::Overhead<etype>>}};
+
+std::unique_ptr<gko::LinOpFactory> generate_solver(
+    const std::shared_ptr<const gko::Executor> &exec,
+    std::shared_ptr<const gko::LinOpFactory> precond,
+    const std::string &description)
+{
+    std::string cb_gmres_prefix("cb_gmres_");
+    if (description.find(cb_gmres_prefix) == 0) {
+        auto s_prec = gko::solver::cb_gmres_storage_precision::keep;
+        const auto spec = description.substr(cb_gmres_prefix.length());
+        if (spec == "keep") {
+            s_prec = gko::solver::cb_gmres_storage_precision::keep;
+        } else if (spec == "reduce1") {
```

Comment:
Technically the first one is unneeded?
```suggestion
        if (spec == "reduce1") {
```

Reply:
We don't need the assignment, but we need the case `keep` because if the suffix does not match any of the choices, it throws a `std::range_error` exception.